Implement workload suggestion API

mhutch opened the pull request:
Adds an API to suggest which workloads should be installed to satisfy a set of missing workload packs.

Fixes the NotImplementException in the workload manifest reader library.

mhutch committed:
Partially implement workload suggestion API for missing packs

mhutch committed:
Finish up workload suggestions API and add tests

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
+
+            // minimize number of unnecessary packs installed by the suggestion, then minimize number of workloads ids listed in the suggestion
+            // eventually a single pass to find the "best" would be more efficient but this this more readable and easier to tweak
+            var scoredList = completeCandidates.Select(
```
nit: use better naming and extract class, extract method to reduce comments.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
+                }
+            }
+
+            //find all valid complete permutations by recursively exploring possible branches from a root
```
nit: same here.

And a method that worth comments every 3 lines is a dangerous method to edit later. We need more test around this method. And it is better to make it a separate method, even a separate class.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
         /// Used by the MSBuild workload resolver to emit actionable errors
         /// </remarks>
-        public IList<WorkloadInfo> GetWorkloadSuggestionForMissingPacks(IList<string> packId)
+        public ISet<WorkloadInfo> GetWorkloadSuggestionForMissingPacks(IList<string> packIds)
```
method too long

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/IWorkloadResolver.cs:
```diff
@@ -9,7 +9,7 @@ public interface IWorkloadResolver
         IEnumerable<WorkloadResolver.PackInfo> GetInstalledWorkloadPacksOfKind(WorkloadPackKind kind);
         IEnumerable<string> GetPacksInWorkload(string workloadId);
-        IList<WorkloadResolver.WorkloadInfo> GetWorkloadSuggestionForMissingPacks(IList<string> packId);
+        ISet<WorkloadResolver.WorkloadInfo> GetWorkloadSuggestionForMissingPacks(IList<string> packId);
```
@dsplaisted please check if the API is what you need

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
+                if (expanded.Any(e => requestedPacks.Contains(e)))
+                {
+                    var stillMissing = new HashSet<WorkloadPackId>(requestedPacks.Where(p => !expanded.Contains(p)));
+                    var satisfied = requestedPacks.Count - stillMissing.Count;
```
not used variable?

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -253,9 +254,161 @@ IEnumerable<WorkloadPackId> ExpandPacks (WorkloadDefinitionId workloadId)
+            var scoredList = completeCandidates.Select(
+                c =>
+                {
+                    var installedCount = c.Packs.Count(p => installedPacks.Contains(p));
```
no obvious test for the scores. This part sounds super complex. If a later change messed up the score logic, a test need to fail.

And it is complicated enough that I cannot say it really works by just reading it.

mhutch committed:
Refactor GetWorkloadSuggestionForMissingPacks for readability

marcpopMSFT commented:
Marked for 5.0.2xx.  The branch isn't open yet but retarget once it is in ~1 month.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:
```diff
@@ -0,0 +1,183 @@
+            /// <summary>
+            /// How many additional (and potentionally unnecessary) packs this suggestion will result in installing
+            /// </summary>
+            public int ExtraPacks { get; internal set; }
```
> How many additional (and potentionally unnecessary)

I don't understand what that means. The number could be too high due to some of the packs are unnecessary? And I need to read the comments to understand what the property does. It is a complex concept, it is fine to have a longer name.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:
```diff
@@ -0,0 +1,183 @@
+        /// <summary>
+        /// A partial or complete suggestion for workloads to install, annotated with which requested packs it does not satisfy
+        /// </summary>
+        class WorkloadSuggestionCandidate : IEquatable<WorkloadSuggestionCandidate>
```
nit: please add explicit modifier per editorconfig

wli3 replied:
I have problem understanding what's this class present. It is a collection of Workloads , Packs but  GatherCompleteSuggestions has a list and a set of them (collection of collection of workload?). What does the branch mean here?

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:
```diff
@@ -0,0 +1,183 @@
+        /// <summary>
+        /// Finds the best value from a list of values according to one or more custom comparators. The comparators are an ordered fallback list - the second comparator is only checked when values are equal according to the first comparator, and so on.
+        /// </summary>
+        private T FindBest<T>(IEnumerable<T> values, params Comparison<T>[] comparators)
```
what's the difference between this and Max() ?

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:
```diff
@@ -0,0 +1,183 @@
+                int hashcode = 0;
+                foreach (var id in Workloads)
+                {
+                    hashcode ^= id.GetHashCode();
```
order matters?

wli3 reviewed (changes requested):
Still don't have enough test cover GatherCompleteSuggestions

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadSuggestionFinder.cs:
```diff
@@ -0,0 +1,183 @@
+                Workloads = id;
+            }
+
+            public HashSet<WorkloadDefinitionId> Workloads { get; }
```
possible to make it immutable?

wli3 commented:
I hope to have some test to cover and document GatherCompleteSuggestions. Example of "recursively explores a branching tree". What kind of candidate is not "the best"

mhutch committed:
Add test for workload suggestion finder

mhutch committed:
Clean up workload pack path aliasing

wli3 commented:
@mhutch could you retarget to release/5.0.2xx? And the logic of union and sort is still hard to understand. We need to have unit tests around it. 1. to document what it does 2. prevent regression and make future change easier, considering it will be someone from SDK team to maintain this code in the long term.

mhutch commented:
@wli3 i've retargeted it. could you be more specific about what you want additional tests and documentation on?

wli3 commented:
@mhutch sorry for being late. I missed the notification somehow. If we could write unit tests for every major methods in WorkloadSuggestionFinder.cs, it would be great. Like FindBest, FilterRedundantSuggestions,GatherCompleteSuggestions, and the constructor of WorkloadSuggestionFinder.

mhutch committed:
Add more tests for WorkloadSuggestionFinder

mhutch commented:
@wli3 The constructor was already implicitly covered by `CanSuggestSimpleWorkload`, `CanSuggestTwoWorkloadsToFulfilTwoRequirements` and `CanSuggestWorkloadThatFulfillsTwoRequirements` and I didn't see a clean way to test it by itself, but I factored out a chunk of logic into `FindPartialSuggestionsAndSimpleCompleteSuggestions` and added a test for that and  `GetBestSuggestion`. I also added some more doc comments, tried to clarify naming, and moved the suggestion finder tests to their own class.

wli3 commented:
Sorry @mhutch I missed the notification of your changes. Please go ahead to merge it after resolving the merge conflict. I can resolve the conflict and merge it if you need.
